runtime/jcl/common/jcldefine.c

```diff
@@ -211,6 +215,13 @@ defineClassCommon(JNIEnv *env, jobject classLoaderObject,
 	}
 
 done:
+	if ((J2SE_VERSION(vm) >= J2SE_19) && (NULL != clazz) && (NULL != hostClass)) {
+		/* From Java 9 and onwards, set IllegalArgumentException when host class and anonymous class have different packages. */
+		if ((clazz != hostClass) && !hasSamePackageName(clazz, hostClass)) {
+			setIllegalArgumentExceptionHostClassAnonClassHaveDifferentPackages(currentThread, clazz, hostClass);
```

Comment:
Preferably the illegal class should not be created, but if that isn't feasible the invalid class should be deallocated. 
Also it seems wrong to just continue after setting the exception to create the local ref and return it. Could set clazz to NULL.

Reply:
Tried to add the error check:
- Before RAM class creation in `defineclass.c::internalDefineClass`: https://github.com/babsingh/openj9/commit/79a3f700c80defd7877b7eaf3d056a37e3a85c84
- During RAM class creation in `createramclass.cpp::internalCreateRAMClassFromROMClassImpl`: https://github.com/babsingh/openj9/commit/85a0411953bdbcaff57e34770eaf238ac8e2c4aa

In both cases, the JVM hung during shutdown. So, it is infeasible to error check without creating the illegal class.

@pshipton Is there existing code to deallocate a J9Class?

Reply:
It would be worth understanding what is going on, it doesn't make sense the check can't be moved elsewhere. Is the same package name check failing during shutdown and causing an exception?

The rom (and ram) memory segments need to be freed via freeMemorySegment(), and the anonymous classloader via freeClassLoader().

Reply:
Maybe with the proper cleanup the hang won't occur.

Reply:
JVM was hanging because I didn't release `vm->classTableMutex` during the error case. I fixed this issue, and pdated pull request with the solution where error-check is done before RAM class.

Native stack for `Unsafe.defineAnonymousClass`:
```
sun_misc_Unsafe.cpp::Java_sun_misc_Unsafe_defineAnonymousClass
   jcldefine.c::defineClassCommon
      defineclass.c::internalDefineClass
         createramclass.cpp:internalCreateRAMClassFromROMClass
            createramclass.cpp::internalCreateRAMClassFromROMClassImpl
```

The error check is added inside `defineclass.c::internalDefineClass` after ROM class creation and before RAM class creation. If error conditions are met, IllegalArgumentException is set, ROM class is freed and `vm->classTableMutex` is released.